### PR TITLE
Make threadcall a bit more threadsafe

### PR DESCRIPTION
**Pull request by gbaraldi**
Threadcall currently assumes that the IO loop gets run in the same thread that spawned it.

While I wish to supercede threadcall by something that uses tasks + adopt_thread. For now lets make it a better.

```julia
julia> function foo()
       Threads.@threads for i = 1:8
           ptr = @threadcall(:jl_malloc, Ptr{Cint}, (Csize_t,), sizeof(Cint))
           @test ptr != C_NULL
           unsafe_store!(ptr, 3)
           @test unsafe_load(ptr) == 3
           ptr = @threadcall(:jl_realloc, Ptr{Cint}, (Ptr{Cint}, Csize_t,), ptr, 2 * sizeof(Cint))
           @test ptr != C_NULL
           unsafe_store!(ptr, 4, 2)
           @test unsafe_load(ptr, 1) == 3
           @test unsafe_load(ptr, 2) == 4
           @threadcall(:jl_free, Cvoid, (Ptr{Cint},), ptr)
       end
       end
foo (generic function with 1 method)

julia> foo()
```
This crashes julia currently

### CODE REVIEW

File: base/threadcall.jl [hunk 1]
```diff
@@ -91,7 +96,12 @@ function do_threadcall(fun_ptr::Ptr{Cvoid}, cfptr::Ptr{Cvoid}, rettype::Type, ar
             fun_ptr, cfptr, args_arr, ret_arr, c_notify_fun, idx)
 
         # wait for a result & return it
-        wait(thread_notifiers[idx])
+        lock(thread_notifiers[idx])
+        try
+            wait(thread_notifiers[idx])
+        finally
+            unlock(thread_notifiers[idx])
+        end
```
Comment (vchuravy) [inline base/threadcall.jl:104]:
```suggestion
        @lock thread_notifiers[idx] begin
            wait(thread_notifiers[idx])
        end
```

File: base/threadcall.jl [hunk 2]
```diff
@@ -1,7 +1,7 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
 const max_ccall_threads = parse(Int, get(ENV, "UV_THREADPOOL_SIZE", "4"))
-const thread_notifiers = Union{Base.Condition, Nothing}[nothing for i in 1:max_ccall_threads]
+const thread_notifiers = Union{Threads.Condition, Nothing}[nothing for i in 1:max_ccall_threads]
```
Comment (vchuravy) [inline base/threadcall.jl:4]: Do we want this to be `ThreadSynchronizer` (e.g. SpinLock backed Condition)?

**Review by vchuravy — APPROVED** [review]

**Comment by gbaraldi** [comment]
This is actually not great because it potentially blocks in the event loop, so I will reimplement using an asyncCondition.

File: base/threadcall.jl (hunk 2 above)
Comment (gbaraldi) [reply base/threadcall.jl:4]: That's the plan. More precisely a Base.Event which now uses a spinlock

**Comment by gbaraldi** [comment]
Bump!

**Review by vchuravy — APPROVED** [review]

**Pull request merged**